### PR TITLE
otelconf: Increased test coverage for Prometheus-related code

### LINKED ISSUE #6821: otelconf: Increase test coverage for Prometheus-related code
Any tests that could be added?

i recall running into issues w/ ipv6 support in the collector when i was manually testing this previously.

_Originally posted by @codeboten in https://github.com/open-telemetry/opentelemetry-go-contrib/pull/6815#pullrequestreview-2628187338_

### PR DESCRIPTION
**Pull request by sonalgaud12**
This PR fixes #6821 

- Added test for prometheusReader function for both v0.3.0 & v0.2.0

- [x] Locally tested

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 ### Added
 
 - Add the `WithPublicEndpoint` and `WithPublicEndpointFn` options to `go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc`. (#7407)
+- Increased test coverage for `prometheusReader` function, including IPv6 and edge-case handling. (#7440)
```
Comment (pellared) [inline CHANGELOG.md]:
```suggestion
```

This does not need a changelog entry.

File: otelconf/v0.2.0/metric_test.go [hunk 2]
```diff
@@ -1156,3 +1156,81 @@ func TestPrometheusIPv6(t *testing.T) {
 		})
 	}
 }
+
+func TestPrometheusReader_ErrorCases(t *testing.T) {
```
Comment (pellared) [inline otelconf/v0.2.0/metric_test.go]: we do not use underscores in test names

```suggestion
func TestPrometheusReaderErrorCases(t *testing.T) {
```

File: otelconf/v0.2.0/metric_test.go [hunk 3]
```diff
@@ -1156,3 +1156,81 @@ func TestPrometheusIPv6(t *testing.T) {
...
+			config: &Prometheus{Host: ptr("localhost")},
+			errMsg: "port must be specified",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			reader, err := prometheusReader(context.Background(), tt.config)
+			assert.Error(t, err)
+			assert.Contains(t, err.Error(), tt.errMsg)
```
Comment (pellared) [inline otelconf/v0.2.0/metric_test.go]: Use https://pkg.go.dev/github.com/stretchr/testify/assert#ErrorContains

File: otelconf/v0.2.0/metric_test.go [hunk 4]
```diff
@@ -1156,3 +1156,81 @@ func TestPrometheusIPv6(t *testing.T) {
...
+	t.Cleanup(func() {
+		require.NoError(t, reader.Shutdown(context.Background()))
+	})
+
+	server := reader.(readerWithServer).server
+	resp, err := http.DefaultClient.Get("http://" + server.Addr + "/metrics")
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		require.NoError(t, resp.Body.Close())
+	})
```
Comment (pellared) [inline otelconf/v0.2.0/metric_test.go]: I think just `defer resp.Body.Close()` would be fine

File: otelconf/v0.2.0/metric_test.go [hunk 5]
```diff
@@ -1156,3 +1156,81 @@ func TestPrometheusIPv6(t *testing.T) {
...
+	reader, err := prometheusReader(context.Background(), cfg)
+	require.NoError(t, err)
+	require.NotNil(t, reader)
+
+	t.Cleanup(func() {
+		require.NoError(t, reader.Shutdown(context.Background()))
+	})
+
+	server := reader.(readerWithServer).server
+	resp, err := http.DefaultClient.Get("http://" + server.Addr + "/metrics")
```
Comment (pellared) [inline otelconf/v0.2.0/metric_test.go]: should we check that `server.Addr` contains `localhost`?

File: otelconf/v0.2.0/metric_test.go [hunk 6]
```diff
@@ -1156,3 +1156,81 @@ func TestPrometheusIPv6(t *testing.T) {
...
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		require.NoError(t, reader.Shutdown(context.Background()))
+	})
+
+	server := reader.(readerWithServer).server
+	assert.Equal(t, 5*time.Second, server.ReadTimeout)
+	assert.Equal(t, 10*time.Second, server.WriteTimeout)
+	assert.Equal(t, 120*time.Second, server.IdleTimeout)
+}
```
Comment (pellared) [inline otelconf/v0.2.0/metric_test.go]: Why are we testing the server timeouts configuration? Is this documented anywhere? If we change then should the test be failing?

File: CHANGELOG.md (hunk 1 above)
Comment (sonalgaud12) [reply CHANGELOG.md]: Removed

File: otelconf/v0.2.0/metric_test.go (hunk 2 above)
Comment (sonalgaud12) [reply otelconf/v0.2.0/metric_test.go]: Fixed

File: otelconf/v0.2.0/metric_test.go (hunk 3 above)
Comment (sonalgaud12) [reply otelconf/v0.2.0/metric_test.go]: Added

File: otelconf/v0.2.0/metric_test.go (hunk 6 above)
Comment (sonalgaud12) [reply otelconf/v0.2.0/metric_test.go]: Removed these assertions

File: otelconf/v0.2.0/metric_test.go (hunk 4 above)
Comment (sonalgaud12) [reply otelconf/v0.2.0/metric_test.go]: Done

File: otelconf/v0.2.0/metric_test.go [hunk 7]
```diff
@@ -1156,3 +1156,145 @@ func TestPrometheusIPv6(t *testing.T) {
...
+		require.NoError(t, reader.Shutdown(context.Background()))
+	})
+
+	server := reader.(readerWithServer).server
+
+	addr := server.Addr
+	assert.True(t, strings.Contains(addr, "localhost") || strings.Contains(addr, "127.0.0.1"),
+		"server address %s should contain localhost or 127.0.0.1", addr)
+
+	resp, err := http.DefaultClient.Get("http://" + server.Addr + "/metrics")
```
Comment (pellared) [inline otelconf/v0.2.0/metric_test.go]:
```suggestion
	resp, err := http.DefaultClient.Get("http://" + addr + "/metrics")
```

File: otelconf/v0.2.0/metric_test.go [hunk 8]
```diff
@@ -1156,3 +1156,145 @@ func TestPrometheusIPv6(t *testing.T) {
...
+				WithResourceConstantLabels: &IncludeExclude{},
+			},
+			errMsg: "binding address",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			reader, err := prometheusReader(context.Background(), tt.config)
+			assert.Error(t, err)
```
Comment (pellared) [inline otelconf/v0.2.0/metric_test.go]: This line is redundant with the next one
```suggestion
```

File: otelconf/v0.2.0/metric_test.go [hunk 9]
```diff
@@ -1156,3 +1156,145 @@ func TestPrometheusIPv6(t *testing.T) {
...
+				WithoutScopeInfo:           ptr(true),
+				WithoutTypeSuffix:          ptr(true),
+				WithoutUnits:               ptr(true),
+				WithResourceConstantLabels: &IncludeExclude{},
+			}
+
+			reader, err := prometheusReader(context.Background(), &cfg)
+			if err != nil {
+				return
+			}
```
Comment (pellared) [inline otelconf/v0.2.0/metric_test.go]: Why is it not?

```go
require.NoError(t, err)
require.NotNil(t, reader)
```

File: otelconf/v0.3.0/metric_test.go [hunk 10]
```diff
@@ -1397,6 +1397,109 @@ func TestPrometheusIPv6(t *testing.T) {
...
+		cfg := Prometheus{
+			Host:                       &host,
+			Port:                       &port,
+			WithoutScopeInfo:           ptr(true),
+			WithoutTypeSuffix:          ptr(true),
+			WithoutUnits:               ptr(true),
+			WithResourceConstantLabels: &IncludeExclude{},
+		}
+		reader, err := prometheusReader(context.Background(), &cfg)
+		assert.Error(t, err)
```
Comment (pellared) [inline otelconf/v0.3.0/metric_test.go]:
```suggestion
```

File: otelconf/v0.3.0/metric_test.go [hunk 11]
```diff
@@ -1397,6 +1397,109 @@ func TestPrometheusIPv6(t *testing.T) {
...
+				WithoutScopeInfo:           ptr(true),
+				WithoutTypeSuffix:          ptr(true),
+				WithoutUnits:               ptr(true),
+				WithResourceConstantLabels: &IncludeExclude{},
+			}
+
+			reader, err := prometheusReader(context.Background(), &cfg)
+			if err != nil {
+				return
+			}
```
Comment (pellared) [inline otelconf/v0.3.0/metric_test.go]: Why is it not?

```go
require.NoError(t, err)
require.NotNil(t, reader)
```

File: otelconf/v0.2.0/metric_test.go [hunk 12]
```diff
@@ -1156,3 +1156,145 @@ func TestPrometheusIPv6(t *testing.T) {
...
+		t.Run(tt.name, func(t *testing.T) {
+			reader, err := prometheusReader(context.Background(), tt.config)
+			assert.Error(t, err)
+			assert.ErrorContains(t, err, tt.errMsg)
+			assert.Nil(t, reader)
+		})
+	}
+}
+
+func TestPrometheusReaderConfigurationOptions(t *testing.T) {
```
Comment (pellared) [inline otelconf/v0.2.0/metric_test.go:1199]: It this test missing for `v0.3.0`?

**Comment by pellared** [comment]
@sonalgaud12, are you able to address the comments?

File: otelconf/v0.2.0/metric_test.go (hunk 9 above)
Comment (sonalgaud12) [reply otelconf/v0.2.0/metric_test.go]: You're right in principle, but require.NoError(t, err) causes the test to fail with "reader is not registered" errors. The if err != nil { return } approach works because it handles cases where the Prometheus reader isn't available in the test environment. Should we keep the working approach or fix the underlying registration issue first?

File: otelconf/v0.2.0/metric_test.go (hunk 12 above)
Comment (sonalgaud12) [reply otelconf/v0.2.0/metric_test.go:1199]: Added

File: otelconf/v0.2.0/metric_test.go (hunk 9 above)
Comment (pellared) [reply otelconf/v0.2.0/metric_test.go]: We should fix the underlying registration issue first as otherwise the test is skipping scenarios where registration does not work. You can also comment out the scenarios which are problematic and add some TODO comment (and create an issue).

Comment (sonalgaud12) [reply otelconf/v0.2.0/metric_test.go]: Added TODO comment and raised a issue

File: otelconf/v0.3.0/metric_test.go [hunk 13]
```diff
@@ -1452,6 +1452,138 @@ func TestPrometheusIPv6(t *testing.T) {
...
+			if len(tt.host) > 2 && tt.host[0] == '[' && tt.host[len(tt.host)-1] == ']' {
+				expectedHost = tt.host[1 : len(tt.host)-1]
+			}
+
+			if expectedHost == "localhost" {
+				assert.True(t, strings.Contains(server.Addr, "localhost") || strings.Contains(server.Addr, "127.0.0.1"),
+					"server address %s should contain localhost or 127.0.0.1", server.Addr)
+			} else {
+				assert.Contains(t, server.Addr, expectedHost)
+			}
```
Comment (pellared) [inline otelconf/v0.3.0/metric_test.go]: Is it possible to not "calculate" the `expectedHost`, but just explicit expected values to `tests` struct e.g. as `wantAddr` field?

File: otelconf/v0.2.0/metric_test.go [hunk 14]
```diff
@@ -1156,3 +1156,139 @@ func TestPrometheusIPv6(t *testing.T) {
...
+
+	reader, err := prometheusReader(context.Background(), cfg)
+	require.NoError(t, err)
+	require.NotNil(t, reader)
+
+	t.Cleanup(func() {
+		require.NoError(t, reader.Shutdown(context.Background()))
+	})
+
+	server := reader.(readerWithServer).server
```
Comment (MrAlias) [inline otelconf/v0.2.0/metric_test.go]: Avoid panicking.

```suggestion
	server, ok := reader.(readerWithServer).server
	require.True(t, ok, "reader is not a readerWithServer")
```

File: otelconf/v0.2.0/metric_test.go [hunk 15]
```diff
@@ -1156,3 +1156,139 @@ func TestPrometheusIPv6(t *testing.T) {
...
+
+	addr := server.Addr
+	assert.True(t, strings.Contains(addr, "localhost") || strings.Contains(addr, "127.0.0.1"),
+		"server address %s should contain localhost or 127.0.0.1", addr)
```
Comment (MrAlias) [inline otelconf/v0.2.0/metric_test.go]: Why should this contain `127.0.0.1` if it was configured with `localhost` as the host?

```suggestion
	assert.Contains(t, addr, "localhost")
```

File: otelconf/v0.2.0/metric_test.go [hunk 16]
```diff
@@ -1156,3 +1156,139 @@ func TestPrometheusIPv6(t *testing.T) {
...
+
+	resp, err := http.DefaultClient.Get("http://" + addr + "/metrics")
```
Comment (MrAlias) [inline otelconf/v0.2.0/metric_test.go]:
```suggestion
	resp, err := http.Get("http://" + addr + "/metrics")
```

File: otelconf/v0.2.0/metric_test.go [hunk 17]
```diff
@@ -1156,3 +1156,139 @@ func TestPrometheusIPv6(t *testing.T) {
...
+
+			reader, err := prometheusReader(context.Background(), &cfg)
+			require.NoError(t, err)
+			require.NotNil(t, reader)
+
+			t.Cleanup(func() {
+				require.NoError(t, reader.Shutdown(context.Background()))
+			})
+
+			server := reader.(readerWithServer).server
```
Comment (MrAlias) [inline otelconf/v0.2.0/metric_test.go]: Same here.

File: otelconf/v0.2.0/metric_test.go [hunk 18]
```diff
@@ -1156,3 +1156,139 @@ func TestPrometheusIPv6(t *testing.T) {
...
+			if len(tt.host) > 2 && tt.host[0] == '[' && tt.host[len(tt.host)-1] == ']' {
+				expectedHost = tt.host[1 : len(tt.host)-1]
+			}
+
+			if expectedHost == "localhost" {
+				assert.True(t, strings.Contains(server.Addr, "localhost") || strings.Contains(server.Addr, "127.0.0.1"),
+					"server address %s should contain localhost or 127.0.0.1", server.Addr)
+			} else {
+				assert.Contains(t, server.Addr, expectedHost)
+			}
```
Comment (MrAlias) [inline otelconf/v0.2.0/metric_test.go]:
```suggestion
			assert.Contains(t, server.Addr, expectedHost)
```

The address should be the one configured. When is this not the case?

File: otelconf/v0.3.0/metric_test.go [hunk 19]
```diff
@@ -1451,6 +1451,135 @@ func TestPrometheusIPv6(t *testing.T) {
...
+
+			t.Cleanup(func() {
+				require.NoError(t, reader.Shutdown(context.Background()))
+			})
+
+			rws, ok := reader.(readerWithServer)
+			require.True(t, ok, "reader is not a readerWithServer")
+			server := rws.server
+
+			assert.NotEmpty(t, server.Addr)
```
Comment (pellared) [inline otelconf/v0.3.0/metric_test.go]: this is not needed (covered already in next line)

```suggestion
```

File: otelconf/v0.2.0/metric_test.go [hunk 20]
```diff
@@ -1155,3 +1155,138 @@ func TestPrometheusIPv6(t *testing.T) {
...
+
+			t.Cleanup(func() {
+				require.NoError(t, reader.Shutdown(context.Background()))
+			})
+
+			rws, ok := reader.(readerWithServer)
+			require.True(t, ok, "reader is not a readerWithServer")
+			server := rws.server
+
+			assert.NotEmpty(t, server.Addr)
```
Comment (pellared) [inline otelconf/v0.2.0/metric_test.go]: this is not needed (covered already in next line)

```suggestion
```

File: otelconf/v0.3.0/metric_test.go [hunk 21]
```diff
@@ -1451,6 +1451,134 @@ func TestPrometheusIPv6(t *testing.T) {
...
+
+	reader, err := prometheusReader(context.Background(), cfg)
+	require.NoError(t, err)
+	require.NotNil(t, reader)
+
+	t.Cleanup(func() {
+		require.NoError(t, reader.Shutdown(context.Background()))
+	})
+
+	server := reader.(readerWithServer).server
```
Comment (MrAlias) [inline otelconf/v0.3.0/metric_test.go]:
```suggestion
	rws, ok := reader.(readerWithServer)
	require.True(t, ok, "reader is not a readerWithServer")
	server := rws.server
```

File: otelconf/v0.3.0/metric_test.go [hunk 22]
```diff
@@ -1451,6 +1451,134 @@ func TestPrometheusIPv6(t *testing.T) {
...
+
+	addr := server.Addr
+	assert.True(t, strings.Contains(addr, "localhost") || strings.Contains(addr, "127.0.0.1"),
+		"server address %s should contain localhost or 127.0.0.1", addr)
```
Comment (MrAlias) [inline otelconf/v0.3.0/metric_test.go]:
```suggestion
	// localhost resolves to 127.0.0.1, so we expect the resolved IP
	assert.Contains(t, addr, "127.0.0.1")
```

File: otelconf/v0.3.0/metric_test.go [hunk 23]
```diff
@@ -1451,6 +1451,136 @@ func TestPrometheusIPv6(t *testing.T) {
...
+
+	rws, ok := reader.(readerWithServer)
+	require.True(t, ok, "reader is not a readerWithServer")
+	server := rws.server
+
+	addr := server.Addr
+	// localhost resolves to 127.0.0.1, so we expect the resolved IP
+	assert.Contains(t, addr, "127.0.0.1")
+
+	resp, err := http.DefaultClient.Get("http://" + addr + "/metrics")
```
Comment (XSAM) [inline otelconf/v0.3.0/metric_test.go]:
```suggestion
	resp, err := http.Get("http://" + addr + "/metrics")
```

nit, as some of your tests use `http.Get`

File: otelconf/v0.3.0/metric_test.go [hunk 24]
```diff
@@ -1451,6 +1451,136 @@ func TestPrometheusIPv6(t *testing.T) {
 	}
 }
 
+func TestPrometheusReaderErrorCases(t *testing.T) {
```
Comment (XSAM) [inline otelconf/v0.3.0/metric_test.go:1455]: Could we use a table-driven approach? Just like the other `TestPrometheusReaderErrorCases` method you wrote?

**Comment by pellared** [comment]
@sonalgaud12, are you able to address @XSAM's comments?

File: otelconf/v0.3.0/metric_test.go (hunk 24 above)
Comment (sonalgaud12) [reply otelconf/v0.3.0/metric_test.go:1455]: Hey, yes
I have updated the code.

**Comment by sonalgaud12** [comment]
@pellared @XSAM I have made all the changes. Can you guys review it once and let me know if anything else is required?

**Comment by sonalgaud12** [comment]
Need approval from @codeboten @MrAlias

**Review by codeboten — APPROVED** [review]
Thanks @sonalgaud12

File: otelconf/v0.3.0/metric_test.go (hunk 24 above)
Comment (sonalgaud12) [reply otelconf/v0.3.0/metric_test.go:1455]: I think it's ready to merge now!
